### PR TITLE
feat(generation): add video generation RL pipeline

### PR DESCRIPTION
**Pull request by chenkang455**
## Summary
This pr focused on adding a video generation (sd3/wan2.2) RL pipeline integrated with Diffusion-NFT, which is the base code for future video policy (vidar) and video WAMs.

To sum up, main modifications include:
- Add SD3 and Wan2.2 generation model wrappers and generation configs.
- Add env interface (OCR, RoboTwin) for the text-guided video generation model.
- Wire generation models into the embodied runner path with NFT actor and rollout support.

Related codebase on diffusion-nft for sd3 and wan2.2 has been fully tested with successful training curves, but the video policy (vidar) part is under development.

```
bash examples/generation/run_generation.sh xxx
```

image

openpi-nft reproduced curve:
image

### CODE REVIEW

File: rlinf/data/datasets/diffusion/__init__.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2026 The RLinf Authors.
```
Comment (thereAreDemonsNearby) [inline rlinf/data/datasets/diffusion/__init__.py:1]: The dataset implementation shouldn't be located in the env directory. RLinf has dedicated data/dataset directory.

File: rlinf/workers/actor/fsdp_nft_policy_worker.py [hunk 2]
```diff
@@ -362,38 +353,94 @@ def nft_forward_and_loss(self, batch):
...
+            gather_type,
+            energy_reduction,
+        )
+        loss_mask = self._align_nft_signal(batch["loss_mask"], e_pos, batch_size)
+        advantages = self._align_nft_signal(batch["advantages"], e_pos, batch_size)
+        advantages = self._postprocess_advantages(advantages)
         # loss
         delta_e = e_pos - e_neg
-        loss = self._compute_nft_loss(e_pos, e_neg, delta_e, advantages, loss_mask)
+        loss_form = self.cfg.algorithm.get("nft_loss_form", "dpo")
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_nft_policy_worker.py]: I recommend encapsulating this part of the loss computation code into _compute_nft_loss, because it's too long

File: rlinf/envs/diffusion/generation_env.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
+# Copyright 2026 The RLinf Authors.
```
Comment (thereAreDemonsNearby) [inline rlinf/envs/diffusion/generation_env.py:1]: It's a bit odd to use the env's interface solely for computing rewards.  We'll discuss this design.

File: rlinf/models/generation/wan/base.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+        raise ValueError(f"Unknown forward_type: {forward_type}")
+
+    def default_forward(self, forward_inputs: dict[str, torch.Tensor], **kwargs):
+        del kwargs
+        batch_size = forward_inputs["nft_x0"].shape[0]
+        num_train_timesteps = min(
+            max(1, int(self.config.num_steps * self.config.timestep_fraction)),
+            self.config.num_steps,
+        )
+        return {
```
Comment (thereAreDemonsNearby) [inline rlinf/models/generation/wan/base.py]: Compared to directly returning all-zero logprobs here, shouldn't we throw a more explicit error?

File: rlinf/models/diffusion/wan/base.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+    num_frames: int = 1
+    num_steps: int = 8
+    eval_num_steps: int = 20
+    timestep_fraction: float = 0.99
+    guidance_scale: float = 1.0
+    eval_guidance_scale: float = 1.0
+    cfg: bool = False
+    max_sequence_length: int = 512
+    output_type: str = "pt"
+    rl_mode: Literal["flow-grpo", "nft"] = "nft"
```
Comment (thereAreDemonsNearby) [inline rlinf/models/diffusion/wan/base.py:41]: Should we add a check at the appropriate place indicating that rl_mode==flow-grpo has not yet been implemented ？

File: rlinf/workers/actor/fsdp_nft_policy_worker.py [hunk 6]
```diff
@@ -272,6 +273,9 @@ def _recompute_v_old(
                 clear_memory()
 
             ref_model = self.model_provider_func()
+            ref_config = getattr(ref_model, "config", None)
+            if hasattr(ref_config, "compile_transformer_forward"):
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_nft_policy_worker.py:279]: please add a comment to explain why compile_transformer_forward is set to false

File: rlinf/workers/actor/fsdp_nft_policy_worker.py [hunk 7]
```diff
@@ -600,12 +673,13 @@ def _compute_nft_weight(
...
-                    .mean(dim=sum_dims, keepdim=True)
-                    .clamp(min=1e-4)
-                    .to(pred.dtype)
+                w = self._reduce_nft_tensor(
+                    torch.abs(pred.double() - target.double()),
+                    gather_type,
+                    "mean",
+                    keepdim=True,
                 )
+                w = w.clamp(min=1e-5).to(pred.dtype)
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_nft_policy_worker.py:772]: minimum for clamp is changed from 1e-4 to 1e-5. Is this necessary ?

File: rlinf/workers/actor/fsdp_nft_policy_worker.py [hunk 8]
```diff
@@ -329,25 +333,12 @@ def nft_forward_and_loss(self, batch):
...
-        # clip delta v and get pos/neg candidates
-        delta_v, clip_coef, v_pos, v_neg = self._compute_clipped_delta_v(
-            v_theta, v_old, sum_dims
-        )
+        v_theta = output_dict["v_theta"][:, :chunk, ...]
+        x_t = forward_inputs["nft_xcur"][:, :chunk, ...]
+        v_old = forward_inputs["nft_v"][:, :chunk, ...].detach()
+        batch_size = x_t.shape[0]
+        # raw delta v and pos/neg candidates
+        delta_v, v_pos, v_neg = self._compute_delta_v_candidates(v_theta, v_old)
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_nft_policy_worker.py:343]: In the old code, the delta_v used for calculating both v_pos and v_neg was clipped. In your new code, you no longer clip delta_v. Is this change safe?

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 9]
```diff
@@ -716,34 +717,40 @@ async def generate_one_epoch(self, input_channel: Channel, output_channel: Chann
...
-                    ),
-                    bootstrap_values=self.get_bootstrap_values(
-                        env_output.get("final_obs", None)
-                    ),
-                    forward_inputs=(
-                        result["forward_inputs"]
-                        if self.rlt_feature_model is not None
-                        else {}
-                    ),
+                    versions=torch.zeros(batch_size, 1, dtype=torch.float32),
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/rollout/hf/huggingface_worker.py]: Around line 600, we have versions = torch.full_like(result["prev_logprobs"]), whose shape is not (batch_size, 1). It seems that writing it this way would raise an error.

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 10]
```diff
@@ -716,34 +717,40 @@ async def generate_one_epoch(self, input_channel: Channel, output_channel: Chann
...
-                rlt_switch_flags=env_output.get("rlt_switch_flags", None),
-                intervene_requested=env_output.get("intervene_flags", None),
-            )
-
-            if self.enable_opd:
-                # OPD keeps this path separate to retain student action tokens for post-rollout teacher logprobs.
-                rollout_result = self._build_rollout_result(
-                    actions,
-                    result,
+            if self.collect_final_values or self.enable_opd:
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/rollout/hf/huggingface_worker.py]: We probably don't need to check whether self.enable_opd is true here. I can't think of a case where enable_opd is true but collect_final_values is false.

File: examples/diffusion/config/vidar_robotwin_nft_action_similarity.yaml [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+algorithm:
+  normalize_advantages: False
+  kl_penalty: kl
+  kl_penalty_type: kl
+
+  group_size: 8
+  reward_coef: 1.0
+  rollout_epoch: 2
+  eval_rollout_epoch: 1
+
```
Comment (thereAreDemonsNearby) [inline examples/diffusion/config/vidar_robotwin_nft_action_similarity.yaml:46]: algorithm.rollout_epoch & algorithm.eval_rollout_epoch is deprecated; use env.train.rollout_epoch & env.train.rollout_epoch instead

File: examples/generation/config/vidar_robotwin_nft_action_similarity.yaml [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+  nft_weight_scale: 1.0
+  nft_target_space: x0
+  nft_xcur_source: resample
+  nft_tau: 1.0
+  nft_clip_ratio: 0.1
+  recompute_v: False
+  nft_gather_type: action_level
+  nft_energy_reduction: mean
+
+  sampling_params:
```
Comment (thereAreDemonsNearby) [inline examples/generation/config/vidar_robotwin_nft_action_similarity.yaml]: algorithm.sampling_params is deprecated for embodied tasks; use rollout.sampling_params instead.

File: examples/generation/config/vidar_robotwin_nft_action_similarity.yaml [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    do_sample: True
+    temperature_train: 1.0
+    temperature_eval: 1.0
+    top_k: 50
+    top_p: 1.0
+    repetition_penalty: 1.0
+    add_BOS: False
+  length_params:
```
Comment (thereAreDemonsNearby) [inline examples/generation/config/vidar_robotwin_nft_action_similarity.yaml]: algorithm.length_params is deprecated for embodied tasks; use rollout.sampling_params.max_new_tokens instead

File: rlinf/workers/actor/fsdp_nft_policy_worker.py (hunk 8 above)
Comment (chenkang455) [reply rlinf/workers/actor/fsdp_nft_policy_worker.py:343]: I found that this mechanism is not working in the previous code, so i just remove it.

File: rlinf/workers/actor/fsdp_nft_policy_worker.py (hunk 7 above)
Comment (chenkang455) [reply rlinf/workers/actor/fsdp_nft_policy_worker.py:772]: This is designed to align with the diffusion-nft, and I have reproduced the openpi-nft code with this worker. Result seems to align with it.

File: rlinf/models/diffusion/wan/base.py (hunk 5 above)
Comment (chenkang455) [reply rlinf/models/diffusion/wan/base.py:41]: Get it

File: examples/diffusion/config/vidar_robotwin_nft_action_similarity.yaml [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    actor: all
+    rollout: all
+    env: all
+
+runner:
+  task_type: embodied
+  logger:
+    log_path: "../results"
+    project_name: rlinf
+    experiment_name: "vidar_rl_action_similarity_video_level_g8_env64_s8_idmbs4_temp2_lr1e5_20260613_221215"
```
Comment (thereAreDemonsNearby) [inline examples/diffusion/config/vidar_robotwin_nft_action_similarity.yaml]: Should these too detailed experiment name be changed? The same issue in wan22 yamls.

File: rlinf/workers/actor/fsdp_nft_policy_worker.py (hunk 8 above)
Comment (thereAreDemonsNearby) [reply rlinf/workers/actor/fsdp_nft_policy_worker.py:343]: Find that the v_pos and v_neg computed here are not used any more. You can remove this method call and  just compute delta_v

Comment (chenkang455) [reply rlinf/workers/actor/fsdp_nft_policy_worker.py:343]:
```
target, pred_pos = self._compute_nft_target_and_pred(
    forward_inputs, target_space, x_t, v_pos, t_bc, dt_bc, sigma_i
)
_, pred_neg = self._compute_nft_target_and_pred(
    forward_inputs, target_space, x_t, v_neg, t_bc, dt_bc, sigma_i
)
```

They are used here? And this function name is reformulated here:

image

Comment (thereAreDemonsNearby) [reply rlinf/workers/actor/fsdp_nft_policy_worker.py:343]: Sorry, I missed some folded code.

**Review by thereAreDemonsNearby — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]
